DOC: Include alternatives to deprecations in the documentation

StefRe opened the pull request:
## PR Summary

There are 90 deprecations currently, for 38 of which alternatives are given in the deprecation warning. These alternatives are not included, however, in the documentation.  
The first commit enables inclusion of alternatives along with links in the documentation and the second commit adds links to the documentation for the existing alternatives if they refer to another function or class.

### Current:
Documentation: ![grafik](https://user-images.githubusercontent.com/19879328/147122069-17decf72-a5d6-470f-a62c-a214f129d6ac.png)

### New: 
Documentation: ![grafik](https://user-images.githubusercontent.com/19879328/147122274-1321e178-6b03-41fd-b1a2-fed0b43bfeee.png)


    Include alternatives to deprecations in the documentation

    - add `alternative` and `addendum` to the documentation by including
      them in the second argument of the `.. deprecated::` directive
    - allow for backticks and optional reference to create links to the
      mentioned alternatives in the documentation
    - remove these additional links and backticks in the deprecation
      warning message that is emitted when using a deprecated item

    Example:
    @_api.deprecated("3.5", alternative="`num2date(e).timestamp() <.num2date>`"
    will create the directive

    .. deprecated:: 3.5
       Use `num2date(e).timestamp() <.num2date>` instead.

    which creates a link to matplotlib.dates.num2date with the caption
    num2date(e).timestamp(). The deprecation warning message will read
    "... Use num2date(e).timestamp() instead."

## PR Checklist

<!-- Please mark any checkboxes that do not apply to this PR as [N/A]. -->
**Tests and Styling**
- [x] Has pytest style unit tests (and `pytest` passes).
- [x] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).

**Documentation**
- [x] New features are documented, with examples if plot related.
- [N/A] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [N/A] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).
- [x] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).

<!--
Thank you so much for your PR!  To help us review your contribution, please
consider the following points:

- A development guide is available at https://matplotlib.org/devdocs/devel/index.html.

- Help with git and github is available at
  https://matplotlib.org/devel/gitwash/development_workflow.html.

- Do not create the PR out of main, but out of a separate branch.

- The PR title should summarize the changes, for example "Raise ValueError on
  non-numeric input to set_xlim".  Avoid non-descriptive titles such as
  "Addresses issue #8576".

- The summary should provide at least 1-2 sentences describing the pull request
  in detail (Why is this change required?  What problem does it solve?) and
  link to any relevant issues.

- If you are contributing fixes to docstrings, please pay attention to
  http://matplotlib.org/devel/documenting_mpl.html#formatting.  In particular,
  note the difference between using single backquotes, double backquotes, and
  asterisks in the markup.

We understand that PRs can sometimes be overwhelming, especially as the
reviews start coming in.  Please let us know if the reviews are unclear or
the recommended next step seems overly demanding, if you would like help in
addressing a reviewer's comments, or if you have been waiting too long to hear
back on your PR.
-->

tacaswell commented:
I think this is 💯 from a functionality point of view, but I am a bit worried about inventing a new link markup scheme and the re-writing to rst.   

Can we just say "this is standard rst that will be rendered" and let the rst leak out to the in-line error message?  If we think that is unacceptable, I would rather add another argument to the deprecation tools to allow what shows up in the documentation to be customized (rather than just dropping the value of *alternative* in).  Text processing is hard in general and in my experience text-processing via regex can be very brittle, accepting a bit of duplication in code that is explicitly going to be deleted is probably less work overall than understanding some (admittedly clever) text mutation rules.

StefRe commented:
Maybe I should have better described the PR as "_Allow rst in `alternative` for documentation and retrieve plain text from that rst for warnings_".

> inventing a new link markup scheme and the re-writing to rst

There is **no new scheme or any re-writing to rst**, everything is standard rst. The `alternative` is just added as is to the second argument of `.. deprecated::` (which used to be just `message`), only the words `Use` and ` instead`  are prepended/appended for readability:
```
        second_arg = ' '.join([t.strip() for t in
                               (message, f"Use {alternative} instead."
                                if alternative else "", addendum) if t])
```
<br>  
<br>  


> text-processing via regex can be very brittle

This is certainly true and I didn't mean to cover all possible corner cases when converting the rst into plain text for the warning. Complex rst expressions with nested and escaped  `` ` ``, `<` and `>` chars or malformed rst expressions will not be correctly handled by the regex, but what will happen: if it's a malformed rst expression, sphinx will issue a warning or an error when building the docs and the error should be fixed. If the rst is valid but too complex for the regex, some garbled text will be included in the warning message and, hopefully, someone will open an issue here (btw, I can't think of any case where such complex rst would be needed for an _alternative_, but it could happen of course; the three standard cases are covered by the test, I could add more exotic cases to the test). There won't be any runtime error or anything affecting the functionality of matplotlib.
<br>  


Possible alternatives:

1. show the rst as-is in the warning
    - if its just a backticked link to another function it's not a big deal, rst where link and caption are different looks a bit strange

2. disallow usage of rst in _alternative_
   - documentation would be plain text only without links (and not to use rst can't be enforced anyway)
   - of course this is better than having no documentation of the _alternative_ at all, but why not using the means provided by sphinx

3. add a new parameter  `alternative_rst` to `_api.deprecated` etc.
   - quick and easy at the cost of writing the *alternative* two times
   - saves two lines of code:    ``if alternative: alternative = re.sub(r"`([^`]+) <.*>`|`\.?([^`]+)`", r"\1\2", alternative)``
   - API change

StefRe commented:
I just saw that the space between the caption and the link is optional (even [not recommended](https://sublime-and-sphinx-guide.readthedocs.io/en/latest/references.html#use-custom-link-text) although the examples in the [rst spec](https://docutils.sourceforge.io/docs/ref/rst/restructuredtext.html#embedded-uris-and-aliases) contain spaces), so I changed the wording in the docs, the regex and the _alternatives_  with separate captions accordingly. I also changed the test a bit to make it easier to read.

tacaswell commented on doc/devel/contributing.rst:
```diff
@@ -326,6 +326,14 @@ Introducing
    the next point release, e.g. "3.x".
 
+   When providing alternatives in ``_api.warn_deprecated()`` or
+   ``@_api.deprecated``, you can use backticks to generate links to these
```
Maybe rephrase this whole paragraph as  "You can use standard rst cross references in the alternative.  The string will be processed to remove back ticks and only include the link caption in the runtime warning"  or something similar?  I think we can expect people who are deprecating functionality to be at least familiar with rst.  I think this is a case where trying to be verbose and clear makes it more confusing ;)

StefRe replied:
Sure, sounds much better. (it's always difficult for me to be concise and clear at the same time when writing in English)

tacaswell commented:
> There is no new scheme or any re-writing to rst,

I misunderstood what was going on 🐑 .  I thought that the transform was on the way into the rst not on the way into the warning text.  I stand by my logic, but it is not relevant to this PR.

----

I do not think option 1 is _terrible_ as even though it might look a bit funny, it does give more information about what is going on (by pointing to the class / method of interest).  We have had some issues in the past where the alternative made perfect sense to us when we reviewed the PR (because we had the context of what class was being edited), it made no sense to users (because the class the method that was suggested as an alternative was on was not clear).  It would be a shame to have that information in the source and documentation, but not visible at run time.

dstansby commented:
This is amazing! I'm very pro overall.

I'm also on the side of option 1 from https://github.com/matplotlib/matplotlib/pull/22038#issuecomment-1000124677. I think the RST formatted text (at least in all the cases and types used in this PR) is parseable as plain text, and the removal of the RST markup isn't worth the hassle of maintaining a rather complex regular expression substitutuion.

StefRe commented:
I removed the rst text processing and pushed it as a separate commit (so it could be reverted easily in case someone complains about strange deprecation warning messages).

QuLogic commented on doc/devel/contributing.rst:
```diff
@@ -326,6 +326,8 @@ Introducing
    All these helpers take a first parameter *since*, which should be set to
    the next point release, e.g. "3.x".
 
+   You can use standard rst cross references in the alternative.
```
```suggestion
   You can use standard rst cross references in *alternative*.
```

StefRe replied:
fixed

QuLogic commented on lib/matplotlib/dates.py:
```diff
@@ -1722,8 +1722,8 @@ def _get_interval(self):
 
-@_api.deprecated("3.5",
-                 alternative="mdates.date2num(datetime.utcfromtimestamp(e))")
+@_api.deprecated("3.5", alternative="" +
+                 "`date2num(datetime.utcfromtimestamp(e))<.date2num>`")
```
```suggestion
@_api.deprecated(
    "3.5",
    alternative="`date2num(datetime.utcfromtimestamp(e))<.date2num>`")
```

StefRe replied:
fixed

QuLogic reviewed (commented):
Probably didn't need to bother with the 3.4 deprecations as they're about to be (or already) delete.d

tacaswell commented:
Thank you @StefRe !

jklymak commented:
OK, so this breaks the docs build if I have:

```
@_api.deprecated(
         "3.6", alternative="figure.get_layout_engine().set(**kwargs)")
    def set_constrained_layout_pads(self, **kwargs):
```

With `WARNING: Inline strong start-string without end-string.`  Anything to be done here?  Could just remove `**kwargs`...

timhoffm commented:
You may try `\*\*kwargs` as a workaround.

jklymak commented:
But that looks terrible in the actual warning?

StefRe commented:
@jklymak I think enclosing it in backticks ```"`figure.get_layout_engine().set(**kwargs)`"```should be enough to silence the warning . This will also neatly format the alternative as code in the docs. If you want link to the `set` method of the layout engine as well, you could write ```"`figure.get_layout_engine().set(**kwargs)<Layout_engine.set>`"``` (given that `get_layout_engine` returns a `Layout_engine` object).

oscargus commented:
@meekseeksdev backport to 3.5.x

oscargus commented:
@meeseeksdev backport to v3.5.x
